### PR TITLE
Add raptor metadata DB optimizations

### LINKED ISSUE #6497: Provide useful clustering for Raptor shard tables
Raptor shard tables should be clustered on something that would be reasonably useful so that queries against these tables have at least some shot of not doing a full-table scan.

I would propose clustering the shard tables on the max value of the temporal column since shards need to be organized along temporal columns. To cluster in MySQL, you need to make to modify the PK, and enforce the original uniqueness with a UNIQUE index.

There's some discussion to be had as to what to cluster on if a table has no temporal column.

### PR DESCRIPTION
**Pull request by erichwang**
Fixes #6497

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 1]
```diff
@@ -179,8 +180,22 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
             }
         }
 
+        ImmutableList.Builder<String> coveringIndexBuilder = ImmutableList.builder();
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]: Use `StringJoiner` for this

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 2]
```diff
@@ -179,8 +180,22 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
...
+        if (temporalColumnId.isPresent()) {
+            // Add the max temporal column first to accelerate queries that scan mostly recent data
+            coveringIndexBuilder.add(maxColumn(temporalColumnId.getAsLong()))
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]: Wrap both `add` calls so they line up

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java [hunk 3]
```diff
@@ -88,14 +89,16 @@
             "  compressed_size BIGINT NOT NULL,\n" +
             "  uncompressed_size BIGINT NOT NULL,\n" +
             "  UNIQUE (shard_uuid),\n" +
+            "  UNIQUE (table_id, bucket_number, shard_id, shard_uuid, create_time, row_count, compressed_size, uncompressed_size),\n" + // Covering index organized by table_id
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java]: Move the comment to the line before the annotation
```java
// include a covering index organized by table_id

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java [hunk 4]
```diff
@@ -88,14 +89,16 @@
...
             "  FOREIGN KEY (table_id) REFERENCES tables (table_id)\n" +
             ")")
     void createTableShards();
 
     @SqlUpdate("CREATE TABLE IF NOT EXISTS shard_nodes (\n" +
             "  shard_id BIGINT NOT NULL,\n" +
             "  node_id INT NOT NULL,\n" +
-            "  PRIMARY KEY (shard_id, node_id),\n" +
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java]: Keep the primary key as-is and add the other index on `(node_id, shard_id)`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java [hunk 5]
```diff
@@ -111,21 +114,25 @@
...
             "  PRIMARY KEY (shard_uuid),\n" +
+            "  UNIQUE (transaction_id, shard_uuid),\n" +
             "  FOREIGN KEY (transaction_id) REFERENCES transactions (transaction_id)\n" +
             ")")
     void createTableCreatedShards();
 
     @SqlUpdate("CREATE TABLE IF NOT EXISTS deleted_shards (\n" +
-            "  shard_uuid BINARY(16) PRIMARY KEY,\n" +
-            "  delete_time DATETIME NOT NULL\n" +
+            "  shard_uuid BINARY(16),\n" +
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java]: Add `NOT NULL` (now that it's not the primary key)

**Comment by electrum** [comment]
Title the second commit "Cluster Raptor index tables on temporal column"

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java [hunk 6]
```diff
@@ -129,9 +129,10 @@
     void createTableCreatedShards();
 
     @SqlUpdate("CREATE TABLE IF NOT EXISTS deleted_shards (\n" +
-            "  shard_uuid BINARY(16) PRIMARY KEY,\n" +
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/SchemaDao.java]: Let's keep the primary key and just add the covering index

**Comment by erichwang** [comment]
ok updated with your comments, back to you @electrum

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 7]
```diff
@@ -179,8 +179,19 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
             }
         }
 
+        StringJoiner coveringIndexColumns = new StringJoiner(", ").setEmptyValue("");
+        if (temporalColumnId.isPresent()) {
+            // Add the max temporal column first to accelerate queries that scan mostly recent data
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]: Phrase this as "queries that usually scan recent data"

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 8]
```diff
@@ -179,8 +179,19 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
             }
         }
 
+        StringJoiner coveringIndexColumns = new StringJoiner(", ").setEmptyValue("");
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]: `setEmptyValue()` is not needed in this case

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 9]
```diff
@@ -179,8 +179,19 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
...
+        if (temporalColumnId.isPresent()) {
+            // Add the max temporal column first to accelerate queries that scan mostly recent data
+            coveringIndexColumns.add(maxColumn(temporalColumnId.getAsLong()));
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]:
```java
temporalColumnId.ifPresent(id -> coveringIndexColumns.add(maxColumn((id))));

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 10]
```diff
@@ -179,8 +179,19 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
...
+            coveringIndexColumns.add(minColumn(temporalColumnId.getAsLong()));
+        }
+
         String sql;
         if (bucketed) {
+            coveringIndexColumns.add("bucket_number")
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]: Put the `.add` on the next line

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java [hunk 11]
```diff
@@ -189,17 +200,25 @@ public void createTable(long tableId, List<ColumnInfo> columns, boolean bucketed
                     tableColumns +
                     "  PRIMARY KEY (bucket_number, shard_uuid),\n" +
                     "  UNIQUE (shard_id),\n" +
-                    "  UNIQUE (shard_uuid)\n" +
+                    "  UNIQUE (shard_uuid),\n" +
+                    "  UNIQUE (" + coveringIndexColumns + ")\n" +
                     ")";
         }
         else {
+            coveringIndexColumns.add("node_ids")
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java]: Same

**Pull request merged**